operator/cilium_node.go

```diff
@@ -20,16 +26,240 @@ import (
+	// KVStore.
+	if cfgBuilder.kvStoreSync {
+		ciliumNodeKVStore, err := store.JoinSharedStore(store.Configuration{
+			Prefix:     nodeStore.NodeStorePrefix,
+			KeyCreator: nodeStore.KeyCreator,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Start a go routine to GC all CiliumNodes from the KVStore that are
+		// no longer running.
+		go func() {
+			<-k8sCiliumNodesCacheSynced
+			// Since we processed all events received from k8s we know that
+			// at this point the list in ciliumNodeStore should be the source of
+			// truth and we need to delete all nodes in the kvNodeStore that are
+			// *not* present in the ciliumNodeStore.
+			listOfCiliumNodes := ciliumNodeStore.ListKeys()
+
+			kvStoreNodes := ciliumNodeKVStore.SharedKeysMap()
+			for _, ciliumNode := range listOfCiliumNodes {
+				// The remaining kvStoreNodes are leftovers that need to be GCed
+				kvStoreNodeName := nodeTypes.GetKeyNodeName(option.Config.ClusterName, ciliumNode)
+				delete(kvStoreNodes, kvStoreNodeName)
+			}
+
+			for _, kvStoreNode := range kvStoreNodes {
+				// Only delete the nodes that belong to our cluster
+				if strings.HasPrefix(kvStoreNode.GetKeyName(), option.Config.ClusterName) {
+					ciliumNodeKVStore.DeleteLocalKey(context.TODO(), kvStoreNode)
+				}
+			}
+		}()
+
+		// nodeManager is nil so we don't need to handle any events managed by
+		// the nodeManager and the events from Kubernetes will only be relevant
+		// for KVStore.
+		if cfgBuilder.nodeManager == nil {
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if ciliumNode := k8s.ObjToCiliumNode(obj); ciliumNode != nil {
+						nodeNew := nodeTypes.ParseCiliumNode(ciliumNode)
+						ciliumNodeKVStore.UpdateKeySync(context.TODO(), &nodeNew)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", obj, obj)
+					}
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					if oldNode := k8s.ObjToCiliumNode(oldObj); oldNode != nil {
+						if newNode := k8s.ObjToCiliumNode(newObj); newNode != nil {
+							if oldNode.DeepEqual(newNode) {
+								return
+							}
+							nodeNew := nodeTypes.ParseCiliumNode(newNode)
+							ciliumNodeKVStore.UpdateKeySync(context.TODO(), &nodeNew)
+						} else {
+							log.Warningf("Unknown CiliumNode object type %T received: %+v", newNode, newNode)
+						}
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", oldNode, oldNode)
+					}
+				},
+				DeleteFunc: func(obj interface{}) {
+					if ciliumNode := k8s.ObjToCiliumNode(obj); ciliumNode != nil {
+						deletedNode := nodeTypes.ParseCiliumNode(ciliumNode)
+						ciliumNodeKVStore.DeleteLocalKey(context.TODO(), &deletedNode)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", obj, obj)
+					}
+				},
+			}
+		} else {
+			// nodeManager not nil thus the events will be processed by
+			// ciliumNodeKVStore and the nodeManager
```

Comment:
There is a lot of code duplication here, would it be better to only have one set of resource event handler funcs that would work both with nil and non-nil nodeManager?

Reply:
There is, I was worried on optimizing the code over readbility. Please see the commit _operator: simplify duplicated code in the event handling_ after this one where I've simplified the code a little bit.